rust/arrow/src/compute/kernels/cast.rs

```diff
@@ -937,114 +1048,169 @@ where
         from.as_any()
             .downcast_ref::<GenericStringArray<Offset>>()
             .unwrap(),
-    )))
+        cast_options,
+    )?))
 }
 
 fn string_to_numeric_cast<T, Offset: StringOffsetSizeTrait>(
     from: &GenericStringArray<Offset>,
-) -> PrimitiveArray<T>
+    cast_options: &CastOptions,
+) -> Result<PrimitiveArray<T>>
 where
     T: ArrowNumericType,
     <T as ArrowPrimitiveType>::Native: lexical_core::FromLexical,
 {
-    let iter = (0..from.len()).map(|i| {
-        if from.is_null(i) {
-            None
-        } else {
-            lexical_core::parse(from.value(i).as_bytes()).ok()
-        }
-    });
+    let vec = (0..from.len())
+        .map(|i| {
+            if from.is_null(i) {
+                Ok(None)
+            } else {
+                let string = from.value(i);
+                let result = lexical_core::parse(string.as_bytes());
+                if cast_options.safe {
+                    Ok(result.ok())
+                } else {
+                    Some(result.map_err(|_| {
+                        ArrowError::CastError(format!(
+                            "Cannot cast string '{}' to value of {} type",
+                            string,
+                            std::any::type_name::<T>()
+                        ))
+                    }))
+                    .transpose()
+                }
+            }
+        })
+        .collect::<Result<Vec<Option<_>>>>()?;
```

Comment:
I wonder if we need the collect here (and the few places below)? It seems like we might be able to use the iter directly for the case where `cast_options.safe == true` 

I worry that unsafe casts may get slower if we always created a vec for them

Reply:
This is a good question. I can test removing them.

Reply:
FYI I tried here: https://github.com/seddonm1/arrow/pull/2 and it seems to work well. I am trying to gather some performance benchmarks to see if it matters